File: src/dyad_app/web_components/copy_to_clipboard.ts
import {LitElement, html} from 'lit';
import {customElement, property} from 'lit/decorators.js';

@customElement('dyad-copy-to-clipboard')
export class CopyToClipboard extends LitElement {
  @property({type: String})
  text = '';

  @property({type: Number})
  counter = 0;

  constructor() {
    super();
    this.text = '';
    this.counter = 0;
  }

  override render() {
    return html``;
  }

  override updated(changedProperties: Map<string, any>) {
    if (changedProperties.has('counter')) {
      this.writeToClipboard(this.text);
    }
  }

  private async writeToClipboard(text: string) {
    if (!text) {
      console.debug('No text to copy');
      return;
    }
    try {
      await navigator.clipboard.writeText(text);
      console.debug('Text successfully copied to clipboard');
    } catch (err) {
      console.error('Failed to copy text: ', err);
    }
  }
}

File: src/dyad_app/web_components/copy_to_clipboard.py
import mesop as me
import mesop.labs as mel
from dyad_app.utils.web_components_utils import get_js_bundle_path


@me.stateclass
class CopyToClipboardState:
    counter: int = 0
    text: str = ""


def write_to_clipboard(text: str):
    state = me.state(CopyToClipboardState)
    state.counter += 1
    state.text = text


@mel.web_component(path=get_js_bundle_path())
def copy_to_clipboard(
    *,
    key: str | None = None,
):
    state = me.state(CopyToClipboardState)
    return mel.insert_web_component(
        name="dyad-copy-to-clipboard",
        key=key,
        properties={
            "text": state.text,
            "counter": state.counter,
        },
    )
